### PR TITLE
:sparkles: add metrics for webhooks

### PR DESCRIPTION
**Pull request by mengqiy**
Add 2 metrics for webhook:
- Number of total requests
- Latency of processing admission requests

### CODE REVIEW

File: pkg/webhook/admission/webhook.go [hunk 1]
```diff
@@ -121,6 +123,8 @@ func (w *Webhook) Handle(ctx context.Context, req atypes.Request) atypes.Respons
 	case types.WebhookTypeValidating:
 		resp = w.handleValidating(ctx, req)
 	default:
+		metrics.FailedRequests.WithLabelValues(w.Name).Inc()
```
Comment (Liujingfang1) [inline pkg/webhook/admission/webhook.go]: I have a general question. Maybe I missed that part. It looks that the failure branches will have metrics. How about the successful ones?

Comment (mengqiy) [reply pkg/webhook/admission/webhook.go]: This is a WIP PR.
I created this PR to see if Travis can reproduce the same strange error I have locally.
I will push more change for the successful ones.

File: pkg/webhook/admission/http.go [hunk 2]
```diff
@@ -71,32 +80,36 @@ func (wh *Webhook) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+func (wh *Webhook) writeResponse(w io.Writer, response types.Response) {
 	encoder := json.NewEncoder(w)
 	responseAdmissionReview := v1beta1.AdmissionReview{
 		Response: response.Response,
 	}
 	err := encoder.Encode(responseAdmissionReview)
 	if err != nil {
 		log.Error(err, "unable to encode the response")
-		writeResponse(w, ErrorResponse(http.StatusInternalServerError, err))
+		metrics.FailedRequests.WithLabelValues(wh.Name).Inc()
```
Comment (DirectXMan12) [inline pkg/webhook/admission/http.go]: why are we incrementing both here and right before we call this method?

File: pkg/webhook/internal/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+			Name: "controller_runtime_webhook_total_requests",
+			Help: "Number of total webhook requests",
+		},
+		[]string{"webhook"},
+	)
+	// SucceededRequests is a prometheus metric which counts the total number of requests that
+	// the webhook server has successfully processed.
+	SucceededRequests = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "controller_runtime_webhook_succeeded_requests",
```
Comment (DirectXMan12) [inline pkg/webhook/internal/metrics/metrics.go]: this should be a label on requests (like `succeeded` as the label, and then `true` and `false` as the values)

File: pkg/webhook/internal/metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	// TotalRequests is a prometheus metric which counts the total number of requests that
+	// the webhook server has received.
+	TotalRequests = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "controller_runtime_webhook_total_requests",
```
Comment (DirectXMan12) [inline pkg/webhook/internal/metrics/metrics.go]: `requests_total`.  By convention, *all* counters end in `_total`.

File: pkg/webhook/internal/metrics/metrics.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+			Name: "controller_runtime_webhook_failed_requests",
+			Help: "Number of total failed webhook requests",
+		},
+		[]string{"webhook"},
+	)
+	// InternalErrorRequests is a prometheus metric which counts the total number of requests that
+	// the webhook server has failed to process with an server internal error.
+	InternalErrorRequests = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "controller_runtime_webhook_internal_error_requests",
```
Comment (DirectXMan12) [inline pkg/webhook/internal/metrics/metrics.go]: again, this should be a label on `requests_total`.

**Comment by DirectXMan12** [comment]
Please make sure you review https://github.com/kubernetes/community/blob/master/contributors/devel/instrumentation.md and https://prometheus.io/docs/practices/naming/ when contributing metrics.

**Comment by mengqiy** [comment]
PTAL
Hope it makes more sense now.

File: pkg/webhook/internal/metrics/metrics.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+		prometheus.CounterOpts{
+			Name: "controller_runtime_webhook_requests_total",
+			Help: "Total number of admission requests",
+		},
+		[]string{"webhook", "succeeded"},
+	)
+
+	// Duration is a prometheus metric which is a histogram of the latency
+	// of processing an admission request.
+	Duration = prometheus.NewHistogramVec(
```
[the remark this answers is not in the document]
Comment (mengqiy) [reply pkg/webhook/internal/metrics/metrics.go]: Renamed

**Comment by mengqiy** [comment]
@DirectXMan12 PTAL

**Comment by shawn-hurley** [comment]
Maybe this is a follow-on PR, but it would be nice to document in the go doc what metrics are registered when using this package?

**Comment by mengqiy** [comment]
Bot seems to be dead :/
Applying labels per https://github.com/kubernetes-sigs/controller-runtime/pull/234#issuecomment-448742843

**Pull request merged**